Fixed a part of the second milestone of the learner dashboard project.

arunabh98 opened the pull request:
This PR includes the completely developed learner dashboard page for our users. The next PR would accomplish 'Sign in to save your progress feature', launching the learner dashboard to our users by asking them to check it out as described in [this](https://docs.google.com/document/d/1bkFz3ci8Jy1zUScgxi442_H8TuHiLe8YkAkF-JrDksk/edit?usp=sharing) document and subscriptions would be removed from the preferences page as they are now visible in the learner dashboard page. 

Here's what the new learner dashboard page looks like -
![image](https://cloud.githubusercontent.com/assets/13691317/26528180/968b3bfe-43c2-11e7-94ea-1efa10126f6a.png)

![image](https://cloud.githubusercontent.com/assets/13691317/26528186/ae836920-43c2-11e7-92c6-b4e35155ac66.png)

![image](https://cloud.githubusercontent.com/assets/13691317/26528411/3582cffa-43c9-11e7-8a9e-6e6c0ff389be.png)

arunabh98 committed:
Part of the second milestone of the Learner Dashboard Project.

arunabh98 committed:
Delete incomplete exp and collections, add other features.

arunabh98 committed:
fix lint errors.

arunabh98 committed:
Further develop learner dashboard page.

arunabh98 committed:
add e2e tests.

arunabh98 committed:
fix merge conflicts.

arunabh98 committed:
Fix a few bugs.

arunabh98 committed:
fix another subscription bug.

seanlip commented on assets/i18n/hi.json:
```diff
@@ -196,6 +197,21 @@
 	"I18N_LANGUAGE_FOOTER_VIEW_IN": "Oppia को देखे:",
+	"I18N_LEARNER_DASHBOARD_COMPLETED_SECTION": "पूर्ण",
+	"I18N_LEARNER_DASHBOARD_COMPLETED_TO_INCOMPLETE_COLLECTIONS": "आपके द्वारा किए गए संग्रह में से <[numberMoved]> को 'प्रगति' अनुभाग में स्थानांतरित किया गया क्योंकि उन्हें नए अन्वेषण जोड़े गए हैं!",
+    "I18N_LEARNER_DASHBOARD_DELETED_COMPLETED_COLLECTION": "आपके द्वारा पूरा किए गए संग्रह में से <[numberDeleted]> को हाल ही में हटा दिया गया है असुविधा के लिए खेद है",
```
Thanks for doing the Hindi translations! Just one note, let's follow the tab/space convention in the rest of this file -- this would comply with the format that's being used by translatewiki.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/learner_dashboard.html:
```diff
@@ -0,0 +1,435 @@
+                 ng-mouseenter="showRemoveIcon=true" ng-mouseleave="showRemoveIcon=false">
+              <img class="remove-icon"
+                   ng-show="showRemoveIcon"
+                   ng-src="<[getStaticImageUrl('/icons/remove.png')]>"
```
Is there any possibility of using a material or fontawesome icon for this, in order to remove the need for an additional server request?

arunabh98 replied:
Done!

seanlip commented on core/controllers/dashboard.py:
```diff
@@ -257,6 +258,204 @@ def _round_average_ratings(rating):
         self.render_json(self.values)
 
 
+class LearnerDashboardPage(base.BaseHandler):
```
The creator and learner dashboards are pretty separate. Would it make sense to split dashboard.py into creator_dashboard.py and learner_dashboard.py? We might want to do something similar for core/templates/dev/head/pages/dashboard.

arunabh98 replied:
I could create a new file by the name learner_dashboard.py and keep the learner dashboard functionality over there. But changing all the function names from `dashboard` to `creator dashboard` seems like a pretty big refactor. Functions by the name dashboard have been used extensively throughout the codebase and even comments - DashboardPage, NotificationsDashboardHandler, DashboardHandler. Instead, can we have a convention like this - just dashboard by default refers to the creator dashboard and those starting with learner refer to the learner dashboard. What do you think of this?

seanlip replied:
I don't think that last convention is a good long-term solution, unfortunately. We used to have just one dashboard before, so "dashboard" was unambiguous; now we have two. So I think we would need to update the naming in any case.

seanlip replied:
(It might be simpler to do the renaming of the existing stuff in a separate PR, though, that gets merged into develop before this one.)

arunabh98 replied:
Done!

arunabh98 committed:
Add last played sort.

seanlip commented on feconf.py:
```diff
@@ -563,6 +564,7 @@ def get_empty_ratings():
 NAV_MODE_DASHBOARD = 'dashboard'
```
Perhaps change to NAV_MODE_CREATOR_DASHBOARD?

arunabh98 replied:
Done!

seanlip commented on core/controllers/dashboard.py:
```diff
@@ -257,6 +258,204 @@ def _round_average_ratings(rating):
+
+        index = 0
+        number_of_deleted_completed_explorations = 0
+        for summary in exp_services.get_exploration_summaries_matching_ids(
```
A lot of this feels somewhat copy-paste -- and it's also way too much logic for the controller layer. Is it possible to abstract some of the commonality out a bit, and also move these to functions in the services layer? The controller layer stuff should basically be: get data from source X, get data from source Y, get data from source Z, assemble into dict, send to frontend (and not much more than that).

arunabh98 replied:
Removed most of the logic from the controller layer.

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -0,0 +1,262 @@
+
+            $scope.remove = function() {
+              /* eslint-disable max-len */
+              if (subSectionName === LEARNER_DASHBOARD_SUB_SECTIONS.EXPLORATIONS) {
```
```
              if (subSectionName ===
                  LEARNER_DASHBOARD_SUB_SECTIONS.EXPLORATIONS) {
```

Might want to shorten the name though. LEARNER_DASHBOARD_SECTIONS perhaps?

arunabh98 replied:
There is already an `LEARNER_DASHBOARD_SECTIONS` which corresponds to the outer menu. This corresponds to the inner menu.

seanlip replied:
Argh, yes, sorry. I don't have a better suggestion...

The only thing I would suggest, though, that is somewhat unrelated to this, is to use SUBSECTION instead of SUB_SECTION, and subsection instead of subSection. "subsection" is one word.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -0,0 +1,262 @@
+            $scope.incompleteCollectionsList.splice(index, 1);
+          }
+        }
+      })
```
Missing semicolon?

arunabh98 replied:
Fixed!

seanlip reviewed (commented):
Hi @Arunabh98 -- I took a very cursory pass, but am going to defer to @jaredsilver and @rachelwchen for review of the code/UI. Thanks!

jaredsilver commented:
Awesome work, @Arunabh98! As result of my losing internet for two days (and therefore losing two days of work time), I'm still a bit behind, so I'll try to review this in-depth tomorrow evening. I apologize for the delay!

arunabh98 committed:
Add pagination. address a part of review comments.

seanlip commented:
Just a quick note on this, btw: we are planning to cut a release on June 3rd. If this gets into develop by then, it can go live on June 10th. (It's also fine if this takes a bit longer, and in fact if it's not ready to ship for any reason then we might want to delay its merge into develop until after the release branch is cut -- I just wanted to let you know these dates for info.)

seanlip commented:
Also, @Arunabh98, I have to say that this is going way faster than I expected! Thanks, and well done on that!

jaredsilver reviewed (changes requested):
This is awesome, @Arunabh98! For now, I just have some really nitpicky design things.

For the empty state:
- [ ] Because the speech bubble is inside the top banner there, the vertically-centered "Learner Dashboard" text looks a bit like it's sinking. I'm not sure about this, but it might look a bit better if it appears centered from the navbar to the top of the speech bubble.

![screen shot 2017-05-30 at 9 55 20 pm](https://cloud.githubusercontent.com/assets/1312199/26613936/e6855934-4582-11e7-9df8-23904533edbe.png)

- [ ] On smaller viewports, the line-height of the text makes a bit tough to discern where one text group ends and the next begins. Perhaps decrease the line-height here, or maybe add some more margin between them?

<img width="529" alt="screen shot 2017-05-30 at 10 00 07 pm" src="https://cloud.githubusercontent.com/assets/1312199/26614048/9170bdde-4583-11e7-926b-30654f027cf7.png">

- [ ] At some viewports, the otter hides behind the dialogue box. (Silly otter!)

<img width="328" alt="screen shot 2017-05-30 at 10 05 06 pm" src="https://cloud.githubusercontent.com/assets/1312199/26614126/1b2ebecc-4584-11e7-8929-1e0315466ce3.png">

- [ ] When "Learner Dashboard" wraps onto a second line, it no longer appears vertically centered.

<img width="345" alt="screen shot 2017-05-30 at 10 06 00 pm" src="https://cloud.githubusercontent.com/assets/1312199/26614145/3f240c56-4584-11e7-8fbb-ab312c5b0a04.png">

jaredsilver commented on assets/i18n/en.json:
```diff
@@ -196,6 +197,22 @@
     "I18N_LANGUAGE_FOOTER_VIEW_IN": "View Oppia in:",
+    "I18N_LEARNER_DASHBOARD_COMPLETED_SECTION": "Completed",
+    "I18N_LEARNER_DASHBOARD_COMPLETED_TO_INCOMPLETE_COLLECTIONS": "<[numberMoved]> of the collections you completed have been moved to the 'in progress' section as new explorations have been added to them!",
+    "I18N_LEARNER_DASHBOARD_DELETED_COMPLETED_COLLECTION": "<[numberDeleted]> of the collections which you completed has been recently deleted. We are sorry for the inconvenience",
```
Add period to the end of the string. Also, can we change `has` to `have` if numberDeleted > 1?

arunabh98 replied:
Actually, there's a sad face that appears at the end of both of these sentences. Also, do I have to add new translations to handle the multiple case?

jaredsilver replied:
Ah, sad face sounds good! Yeah, I think we should have a separate string for one versus more than one collection because different languages handle the associated grammar differently.

arunabh98 replied:
Done!

jaredsilver commented on assets/i18n/en.json:
```diff
@@ -196,6 +197,22 @@
+    "I18N_LEARNER_DASHBOARD_COMPLETED_SECTION": "Completed",
+    "I18N_LEARNER_DASHBOARD_COMPLETED_TO_INCOMPLETE_COLLECTIONS": "<[numberMoved]> of the collections you completed have been moved to the 'in progress' section as new explorations have been added to them!",
+    "I18N_LEARNER_DASHBOARD_DELETED_COMPLETED_COLLECTION": "<[numberDeleted]> of the collections which you completed has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_DELETED_COMPLETED_EXPLORATION": "<[numberDeleted]> of the explorations which you completed has been recently deleted. We are sorry for the inconvenience",
```
Add period to the end of the string. Also, can we change `has` to `have` if numberDeleted > 1?

arunabh98 replied:
Actually, there's a sad face that appears at the end of both of these sentences. Also, do I have to add new translations to handle the multiple case?

jaredsilver replied:
Ah, sad face sounds good! Yeah, I think we should have a separate string for one versus more than one collection because different languages handle the associated grammar differently.

arunabh98 replied:
Done!

jaredsilver commented on assets/i18n/en.json:
```diff
@@ -196,6 +197,22 @@
+    "I18N_LEARNER_DASHBOARD_COMPLETED_TO_INCOMPLETE_COLLECTIONS": "<[numberMoved]> of the collections you completed have been moved to the 'in progress' section as new explorations have been added to them!",
+    "I18N_LEARNER_DASHBOARD_DELETED_COMPLETED_COLLECTION": "<[numberDeleted]> of the collections which you completed has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_DELETED_COMPLETED_EXPLORATION": "<[numberDeleted]> of the explorations which you completed has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_COLLECTION": "<[numberDeleted]> of the collections in progress has been recently deleted. We are sorry for the inconvenience",
```
Add period to the end of the string. Also, can we change `has` to `have` if numberDeleted > 1?

arunabh98 replied:
Done!

jaredsilver commented on assets/i18n/en.json:
```diff
@@ -196,6 +197,22 @@
+    "I18N_LEARNER_DASHBOARD_DELETED_COMPLETED_COLLECTION": "<[numberDeleted]> of the collections which you completed has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_DELETED_COMPLETED_EXPLORATION": "<[numberDeleted]> of the explorations which you completed has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_COLLECTION": "<[numberDeleted]> of the collections in progress has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_EXPLORATION": "<[numberDeleted]> of the explorations in progress has been recently deleted. We are sorry for the inconvenience",
```
Add period to the end of the string. Also, can we change `has` to `have` if numberDeleted > 1?

arunabh98 replied:
Done!

jaredsilver commented on assets/i18n/en.json:
```diff
@@ -196,6 +197,22 @@
+    "I18N_LEARNER_DASHBOARD_DELETED_COMPLETED_EXPLORATION": "<[numberDeleted]> of the explorations which you completed has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_COLLECTION": "<[numberDeleted]> of the collections in progress has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_EXPLORATION": "<[numberDeleted]> of the explorations in progress has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_EMPTY_COMPLETED_COLLECTIONS": "It looks like you haven't completed any collection yet. Head over to the library to start an exciting new collection!",
```
>It looks like you haven't completed any collections yet. Head over to the library to start an exciting new collection!

arunabh98 replied:
Done!

jaredsilver commented on assets/i18n/en.json:
```diff
@@ -196,6 +197,22 @@
+    "I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_COLLECTION": "<[numberDeleted]> of the collections in progress has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_EXPLORATION": "<[numberDeleted]> of the explorations in progress has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_EMPTY_COMPLETED_COLLECTIONS": "It looks like you haven't completed any collection yet. Head over to the library to start an exciting new collection!",
+    "I18N_LEARNER_DASHBOARD_EMPTY_COMPLETED_EXPLORATIONS": "It looks like you haven't completed any exploration yet. Head over to the library to start an exciting new exploration!",
```
>It looks like you haven't completed any explorations yet. Head over to the library to start an exciting new exploration!

arunabh98 replied:
Done!

jaredsilver commented on assets/i18n/en.json:
```diff
@@ -196,6 +197,22 @@
+    "I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_EXPLORATION": "<[numberDeleted]> of the explorations in progress has been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_EMPTY_COMPLETED_COLLECTIONS": "It looks like you haven't completed any collection yet. Head over to the library to start an exciting new collection!",
+    "I18N_LEARNER_DASHBOARD_EMPTY_COMPLETED_EXPLORATIONS": "It looks like you haven't completed any exploration yet. Head over to the library to start an exciting new exploration!",
+    "I18N_LEARNER_DASHBOARD_EMPTY_INCOMPLETE_COLLECTIONS": "It looks like you don't have any collections in progress. Head over to the library to start an exciting new collection!",
```
>It looks like you haven't started any collections yet. Head over to the library to start an exciting new collection!

arunabh98 replied:
Done!

jaredsilver commented on assets/i18n/en.json:
```diff
@@ -196,6 +197,22 @@
+    "I18N_LEARNER_DASHBOARD_EMPTY_COMPLETED_COLLECTIONS": "It looks like you haven't completed any collection yet. Head over to the library to start an exciting new collection!",
+    "I18N_LEARNER_DASHBOARD_EMPTY_COMPLETED_EXPLORATIONS": "It looks like you haven't completed any exploration yet. Head over to the library to start an exciting new exploration!",
+    "I18N_LEARNER_DASHBOARD_EMPTY_INCOMPLETE_COLLECTIONS": "It looks like you don't have any collections in progress. Head over to the library to start an exciting new collection!",
+    "I18N_LEARNER_DASHBOARD_EMPTY_INCOMPLETE_EXPLORATIONS": "It looks like you don't have any explorations in progress. Head over to the library to start an exciting new exploration!",
```
>It looks like you haven't started any explorations yet. Head over to the library to start an exciting new exploration!

arunabh98 replied:
Done!

jaredsilver commented on assets/i18n/en.json:
```diff
@@ -196,6 +197,22 @@
+    "I18N_LEARNER_DASHBOARD_EMPTY_COMPLETED_EXPLORATIONS": "It looks like you haven't completed any exploration yet. Head over to the library to start an exciting new exploration!",
+    "I18N_LEARNER_DASHBOARD_EMPTY_INCOMPLETE_COLLECTIONS": "It looks like you don't have any collections in progress. Head over to the library to start an exciting new collection!",
+    "I18N_LEARNER_DASHBOARD_EMPTY_INCOMPLETE_EXPLORATIONS": "It looks like you don't have any explorations in progress. Head over to the library to start an exciting new exploration!",
+    "I18N_LEARNER_DASHBOARD_EMPTY_SUBSCRIPTIONS": "It looks like you have not subscribed to any creator yet. Head over to the library to discover new creators and their amazing explorations!",
```
>It looks like you haven't subscribed to any creators yet. Head over to the library to discover new creators and their amazing explorations!

arunabh98 replied:
Done!

jaredsilver reviewed (changes requested):
A couple copy-related nits. This is all I have time for at the moment, but I can look more in depth over the weekend.

Also, how did you go about marking explorations as complete and in-progress if that functionality isn't yet implemented?

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -0,0 +1,328 @@
+            $scope.remove = function() {
+              /* eslint-disable max-len */
+              if (subSectionName === LEARNER_DASHBOARD_SUB_SECTIONS.EXPLORATIONS) {
+              /*eslint-enable */
```
/* eslint-enable max-len */

(we generally do these things in pairs)

arunabh98 replied:
Fixed.

arunabh98 committed:
address review comments.

arunabh98 committed:
Address more review comments.

arunabh98 committed:
fix merge conflicts.

arunabh98 committed:
address review comments.

arunabh98 committed:
remove mascot image.

arunabh98 commented:
@jaredsilver I have fixed all of the front end errors you mentioned.

arunabh98 commented:
This is ready for another review.

seanlip commented on core/controllers/learner_dashboard.py:
```diff
@@ -0,0 +1,206 @@
+        if self.user_id is None:
+            raise self.PageNotFoundException
+
+        incomplete_exploration_ids = (
```
Um, sorry, no -- still too much logic. The only thing that should be here is:

    incomplete_exp_summaries, num_deleted_incomplete_exps = (
        learner_progress_services.get_incomplete_exp_summaries(self.user_id))

and all the other stuff should go in services. Ditto below.

It also worries me a bit that GET is state-changing in this case, but I can't see a better way to handle this so let's leave it as is. At least it's always an authenticated GET.

Speaking of that, though -- you'll need a `@base.require_user` decorator on this handler, right? Seems better to me to have the redirect than showing a 404.

arunabh98 replied:
Moved this logic to the backend. Also, the learner dashboard page has `@base.require_user` so we do redirect the learners. The data handler controller right? This is modeled after the creator dashboard handler method.

seanlip replied:
Ah, my bad, sorry. Yep, you're right about that bit!

arunabh98 replied:
Done!

seanlip commented on core/controllers/learner_dashboard.py:
```diff
@@ -0,0 +1,206 @@
+        completed_collection_summary_dicts = []
+        incomplete_collection_summary_dicts = []
+
+        for collection_summary in completed_collection_summaries:
```
There's still lots of logic and manual assembly here. I suggest finding a way to push it back into the services layer so that you can just do something like get_completed_collection_summary_dicts(). Alternatively you can also do, in the final dict:

`'completed_collections_list': [summary.to_dict() for summary in completed_collection_summaries]`

which is probably preferable -- it makes sense for the to_dict() method to be on the summary object. Ditto below.

seanlip commented on core/controllers/learner_dashboard.py:
```diff
@@ -0,0 +1,206 @@
+            subscription_list.append(subscription_summary)
+
+        self.values.update({
+            'completed_explorations_list': completed_exp_summary_dicts,
```
In one GET call, how many datastore queries are you making? You might want to check the waterfall graph on /_ah/stats for a learner with, say, 50 explorations.

I'm concerned that we'll run into a similar situation like we did with the creator dashboard where it ended up taking over 1 minute to load for creators with about 60 explorations. We might want to find a way to batch datastore calls using get_multi() if possible.

arunabh98 replied:
Could you explain this a bit more, please? I am having trouble understanding this.

seanlip replied:
In order to reduce network latency it's important to try and make fewer calls from the Oppia server to the datastore. A common technique to do this is to fetch multiple entities from the datastore in one RPC.

By going to /_ah/stats on your local server you can also see a "waterfall graph" that should help you see how many calls are being made. See the [AppStats](https://cloud.google.com/appengine/docs/standard/python/tools/appstats) reference for more info.

Does this help? You can also look out for usages of ndb.get_multi() and ndb.put_multi() in the codebase.

seanlip replied:
Btw, note that you might need to test this with several explorations to see the effects in practice. It might be worth collaborating with @shaz13 and @souravbadami on #2508 if there are utility functions in the admin dashboard for this that you'd all find useful. In the meanwhile, though, it should be possible to batch the datastore calls.

arunabh98 replied:
I had a look at the waterfall graph. I have a small doubt - I am already fetching all the summaries for all the explorations using get_multi which as I understand fetches the summaries in a single call rather than multiple calls. So what exactly do you mean by batch datastore calls using get_multi? Do you mean fetching a fixed number of explorations? Like 16 at a time?

seanlip replied:
Ah, sorry, no. What I forgot to point out was that ndb.get_multi() allows you to fetch multiple entities at a time even though they're of different types. See https://cloud.google.com/appengine/docs/standard/python/ndb/creating-entities#Python_Batch_operations

In your PR you are doing individual gets for each of the "get info about user progress" components, right? The point I was making was that perhaps these could be combined. So maybe you need just one (or two) datastore GET calls at the outset to get all the entities you need first, and then you can process all the domain objects as needed.

seanlip commented on core/domain/learner_progress_services.py:
```diff
@@ -420,3 +470,46 @@ def get_all_incomplete_collection_ids(user_id):
+
+
+def remove_none_entities_from_list(
+        user_id, entity_summary_list, entity_ids, section, sub_section):
```
sub_section --> subsection

I feel somewhat iffy about this function; it has serious side-effects that are not conveyed well in its name. Perhaps consider having a single master method (that handles both completed and in-progress explorations and collections) that does something like this instead:

    def update_and_retrieve_user_progress(user_id):
        # Bulk-retrieve all the user progress data (using ndb.get_multi())
        # If needed, process removals and bulk-put() any models which have changed
        # Return up-to-date user progress data

Then you get two datastore calls per-get, which seems reasonable. Not sure how many you have currently...

arunabh98 replied:
Method no longer needed.

seanlip commented on core/domain/learner_progress_services.py:
```diff
@@ -420,3 +470,46 @@ def get_all_incomplete_collection_ids(user_id):
+        if summary is None:
+            number_removed = number_removed + 1
+            if section == 'complete':
+                if sub_section == 'explorations':
```
Just as a side note, these should really be [named constants](http://wiki.c2.com/?NamedConstants), not "magic strings". This might be obviated by the proposed refactor (see other comments) but I wanted to point it out for future reference.

arunabh98 replied:
Will remember this!

seanlip reviewed (changes requested):
Hi @Arunabh98! Just took a look at the backend; I have  some concerns about speed and scalability with the current approach. Deferring a review of the frontend to the other reviewers.

arunabh98 committed:
address review comments 2.

arunabh98 committed:
Push collection summary dict logic to services layer.

arunabh98 committed:
fix lint errors.

jaredsilver reviewed (changes requested):
Awesome, those changes look great. Thanks @Arunabh98!

A couple more things:

- [ ] At small viewports, these labels look a little weird because of the line height. Also, it would be good for those numbers to align vertically even though the first two have labels that take up two lines and the third does not.

<img width="379" alt="screen shot 2017-06-04 at 8 39 44 pm" src="https://cloud.githubusercontent.com/assets/1312199/26767673/c1e9ca7c-4967-11e7-90f3-95f0f02cecec.png">

- [ ] If I create an exploration with two continue buttons and an end button, clicking through one continue button does not seem to result in the exploration displaying in my 'in progress' section. Should we modify this so that clicking through continue buttons at the beginning of an exploration would be enough for the exploration to be counted as in progress? (We could also file this for a separate issue if it requires changing a substantial amount of logic.)

- [ ] At certain viewports, the caret in the left menu appears on its own line, and the width of the header gets really small resulting in word wrapping even though there's enough room for it all to be on one line.

<img width="883" alt="screen shot 2017-06-04 at 8 51 41 pm" src="https://cloud.githubusercontent.com/assets/1312199/26767670/b7a168a4-4967-11e7-904e-84a6a6f9d85b.png">

- [ ] At certain viewports, there is no whitespace between the label text and the vertical lines.

<img width="650" alt="screen shot 2017-06-04 at 9 04 02 pm" src="https://cloud.githubusercontent.com/assets/1312199/26767886/741771da-4969-11e7-949b-34fe4c2eca93.png">

arunabh98 commented:
@jaredsilver 
1) Fixed the line height issue and made sure the numbers are aligned with each other.
2) I think there is been a mistake. If I click the first continue button and move to the next state and if we then leave the exploration in the middle the exploration is added to the in progress section. If the learner completes it then it is added to the completed section. However, if the learner has not left the exploration, that is the exploration is open in a tab that means the learner is still on it. If he/she leaves it we mark it as in progress so that the learner return back and complete the exploration.
3) Fixed!
4) Fixed!

arunabh98 committed:
address review comments 3.

jaredsilver commented:
@Arunabh98 -- RE point 2, you're right! Sorry about that, and thanks for the clarification!

I think this looks fantastic, and I haven't seen any bugs from running through it.

One very small nit is perhaps the carets should rotate depending on whether the accordion menu is open or closed for a section, but that's not a huge deal.

I didn't test every permutation of {exploration|collection} being {edited|deleted} when {in progress|completed}, but everything else seems good to go on the front end.

Awesome work, @Arunabh98!
